### ISSUE TITLE
OpenWeatherMap failed to retrieve data, onecall version 3.0 not supported (for new OWM accounts)

### ISSUE THREAD

**Issue by gsagsa99** [domoticz/domoticz#5336]
Domoticz Version: 2022.1 (build 14477)
Platform : Windows 10

OpenWeatherMap is not supporting version 2.5 of the 'onecall' API.  

I specified the city ID = 6455259 (Paris).  The first call (version 2.5) worked fine .  The pair Lat/Lon is retrieved.
This pair is used to build version 2.5 of 'onecall' API.   This is failing.

Below the log

2022-09-10 11:08:24.727  Debug: Meteo_Local: Got location parameter 6455259
2022-09-10 11:08:24.727  Debug: Meteo_Local: Starting with setting 0, 0, 0, 0
2022-09-10 11:08:24.727  Status: Meteo_Local: Using specified location (Station ID 6455259)!
2022-09-10 11:08:24.728  Debug: Meteo_Local: Get data from https://api.openweathermap.org/data/2.5/weather?id=6455259&appid=f4bc1306c8c0bda6307d0dc8941437a4&units=metric&lang=fr
2022-09-10 11:08:24.903  Status: Meteo_Local: City (6455259)-> Lat/Long = 48.8565,2.3524
2022-09-10 11:08:24.904  Status: Meteo_Local: Started
2022-09-10 11:08:27.907  Debug: Meteo_Local: Get data from https://api.openweathermap.org/data/2.5/onecall?lat=48.8565&lon=2.3524&exclude=minutely&appid=f4bc1306c8c0bda6307d0dc8941437a4&units=metric&lang=fr
2022-09-10 11:08:28.005  Debug: HTTP 401: Unauthorized
2022-09-10 11:08:28.007  Error: Meteo_Local: Error getting http data!

I checked both using a browser :  the first one (weather) is OK and the second one fails.

Instead using the version 3.0 of 'onecall' is OK.  Below the working URL (**just replace 2.5 with 3.0**)

https://api.openweathermap.org/data/3.0/onecall?lat=48.8565&lon=2.3524&exclude=minutely&appid=f4bc1306c8c0bda6307d0dc8941437a4&units=metric&lang=fr

Regards
gsagsa99

**Comment by kiddigital** [comment]
@gsagsa99 , thx for reporting. I will take a look asap...

Did not get/see any deprecation message/e-mail coming by. Also no message on their website as far as I can see.

From 2.5 to 3.0 might be (assuming semantic versioning) a backward compatibility breaking change... hopefully not too much has changed. Would be great if we can find some documentation/changelog revealing the API changes.

**Comment by gsagsa99** [comment]
Thanks @kiddigital 
Apparently the 2.5 is deprecated for new account .  I created my account and got my key one week ago
Found this on the Net 
https://github.com/openhab/openhab-addons/issues/13314
https://github.com/openhab/openhab-addons/issues/13374

**Comment by kiddigital** [comment]
Ah.. ok... thx for this info.. Was just looking at my existing OWM hardware module in Domoticz and that one is working fine.

Also when using my existing API key on a test Domoticz environment with a newly setup OWM module, it works as expected.

So it is most likely as described in the OpenHab issue, newly/recently created OWM API keys only work with the 3.0. I will try to see what happens if we just replace the 2.5 in 3.0... maybe it just works...

**Comment by kiddigital** [comment]
Did some quick testing... not conclusive yet...

If I switch to use the OneCall 3.0 endpoint, I get a `401: Unauthorized` message. I am using the (old) API key that does work when calling the old (2.5) endpoint.

To test I generated a new API key, but that still gives me the 401. But this is likely caused by the API key not being active yet (it give 401 first at the 2.5 weather calls, but these calls do work now. So just have to be patient I think).
  
At the moment, we _abuse_ the weather data call to do the conversion between Location(s) and its Latitude/Longitude. Maybe it is better to switch to the dedicated [GeoCoding API](https://openweathermap.org/api/geocoding-api) for that purpose.

**Comment by gsagsa99** [comment]
Thanks for the feedback.

Indeed I saw the new GeoCoding API.

Also I think tha OWM not only mandate to open an account to get an API key
but mandate to have a 'Base Plan'.

Without 'Base Plan' , only the 'Weather' API is answered

To get 'onecall' API working, you need to subscribe a base plan even if the
first 1000 per day are free !

Regards
gsagsa99


[image: width=]
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

Le lun. 12 sept. 2022 à 21:33, KidDigital ***@***.***> a
écrit :

**Comment by gsagsa99** [comment]
Hi

https://openweathermap.org/price#weather



[image: width=]
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

Le lun. 12 sept. 2022 à 21:33, KidDigital ***@***.***> a
écrit :

**Comment by kiddigital** [comment]
Hm... can you try to get an API key for the _'free'_ plan? And check if that one works?

This _'base'_ plan for the OneCall API seems to be new and previously the OneCall API (before 3? Like the 2.5?) was included as one of the API's available in the _free_ plan (next to the weather, geoCoding, etc. API's).

**Comment by gsagsa99** [comment]
Hello

I have TWO accounts now.

1.  A free account with a subscription of a Base PLAN (first 1000 calls are
free and billing after)
2.  A free acount NO SUBSCRIPTION

Hereafter the tests and the results

*Free account WITH SUBSCRIPTION*

WEATHER -
https://api.openweathermap.org/data/2.5/weather?lat=50.7001&lon=4.988987&appid=f4bc1306c8c0bda6307d0dc8941437a4&units=metric&lang=fr
--> *OK*

GEO -
http://api.openweathermap.org/geo/1.0/direct?q=ENINES,BE}&limit=1&appid=f4bc1306c8c0bda6307d0dc8941437a4
-- *OK*

ONECALL -
https://api.openweathermap.org/data/2.5/onecall?lat=50.7001&lon=4.988987&appid=f4bc1306c8c0bda6307d0dc8941437a4&units=metric&lang=fr
--> *FAILED*

ONECALL -
https://api.openweathermap.org/data/3.0/onecall?lat=50.7001&lon=4.988987&appid=f4bc1306c8c0bda6307d0dc8941437a4&units=metric&lang=fr
--> *OK*

The results make sense as it is a new account the onecall 2.5 is deprecated.
Please not the the Weather 2.5 will be deprecated soon and the GEO must be
used

*Free account NO SUBSCRIPTION*

WEATHER -
https://api.openweathermap.org/data/2.5/weather?lat=50.7001&lon=4.988987&appid=364d01e717bf215a28cc072aed7234cd&units=metric&lang=fr
--> *OK*

GEO -
http://api.openweathermap.org/geo/1.0/direct?q=ENINES,BE}&limit=1&appid=364d01e717bf215a28cc072aed7234cd
--> *OK*

ONECALL -
https://api.openweathermap.org/data/2.5/onecall?lat=50.7001&lon=4.988987&appid=364d01e717bf215a28cc072aed7234cd&units=metric&lang=fr
--> *FAILED*

ONECALL -
https://api.openweathermap.org/data/3.0/onecall?lat=50.7001&lon=4.988987&appid=364d01e717bf215a28cc072aed7234cd&units=metric&lang=fr
--> *FAILED*

As it is a new account, the onecall 3.0 failed.  In theory , for an old
account , the onecall 2.5 should work for backward compatibility.  But I
could not test it.

Hope that help

if needed dot testing , you could use my keys as long as you stay in the
free limit :-)

Regards
Gsagsa99












[image: width=]
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

Le lun. 12 sept. 2022 à 22:04, KidDigital ***@***.***> a
écrit :

**Comment by kiddigital** [comment]
Thx again @gsagsa99 , could you try to add the _free_ plan to the account with NO subscription? (Or use a 3rd account) and let me know what that does?

**Comment by gsagsa99** [comment]
Hi @kiddigital , 
I am not sure I understand what you are asking.

I have two accounts   both are free plans.  One has a subscription , the other NO
I did the tests for both.

I am not sure what you are asking ?
Regards

**Comment by kiddigital** [comment]
Hello @gsagsa99 , maybe we mean the same but call it different.

Currently I have the following _billing plans_
![image](https://user-images.githubusercontent.com/12084477/190105168-071d905c-33f6-41b8-8248-c87cea71aa7b.png)

And the following _Services_
![image](https://user-images.githubusercontent.com/12084477/190105663-8df6f161-0dcc-41a2-b57e-c3de316d1d3e.png)

When looking at the _API keys_ tab,  I have 2 API keys. My old one, generated two year ago or so. And a new one generated a few days ago. But unfortunately, it is not clear which API key works for which API (and version).

I was not sure if your second account (without the OneCall subscription) has which _billing plans_ and _Services_. Hence my question.

To get a better understanding, I have reached out to their support desk. Let's see what they can tell us.

Sorry that for your situation it does not work yet. If I would change the module now (switch to use the 3.0 OneCall), it will break for the current users.

**Comment by gsagsa99** [comment]
Hi @kiddigital 
Based on my understanding:
1. your old key will still work with the version 2.5 call. This is because your key is old (two years)
2. your new key , I am not sure.  Just try manually with a browser to get the answer


I my case , I have two accounts I just created a few days ago
1. First account with a Billing plan (called Base Plan), one key 
       Version 2.5 does not work , version 3.0 works
2. Second account no billing plan , one key
       None is working

Thanks

**Comment by kiddigital** [comment]
> I my case , I have two accounts I just created a few days ago
> 
> 1. First account with a Billing plan (called Base Plan), one key 
> 
>        Version 2.5 does not work , version 3.0 works
> 
> 2. Second account no billing plan , one key
> 
>        None is working
> 

Is it possible to have your 2nd account to subscribe to the 'free' plan? (See my screenshot above).

It seems logical that if your 2nd account isn't subscribed to any thing at all, the API key will not provide access to any data.

**Comment by gsagsa99** [comment]
Hi @kiddigital 

The free plan you are talking is a professional one and is the default.  I think there is nothing to subscribe.

I used my account (no subscription) and I could call the following API

https://api.openweathermap.org/data/2.5/forecast?lat=50.7001&lon=4.988987&appid=364d01e717bf215a28cc072aed7234cd&units=metric&lang=fr

This works fine but I don't know if the json output is the same .

It is not the '**onecall**' but for professional it is '**forecast**'.

It is a version 2.5.    The 3.0 does not exist

regards

**Comment by kiddigital** [comment]
Ok, clear. We do mean the same indeed.

So now we have a decision to make for the future (but I will also await input from OWM if/once they respond to my question):
- In the 'old' situation: the __free__ _professional_ subscription (= default subscription) gave access to many API's _including_ the __OneCall__ API
- Now in the 'new' situation: the same __free__ _professional_ subscription still gives access to many API's, but __NOT__ anymore to the __OneCall__ API. The _OneCall_ API now has its own subscription

The current Domoticz OWM modules leverages 2 API's, _weather_ (but for the 'wrong' reason, it is used to retrieve Lon/Lat from a name. Which should now be done using the _Geo_ API!) and _OneCall_. These 2 API's now need different (or both?) subscriptions.

We should make a choice which (or both) subscription to use for the OWM module.

Looking at what we do/need and how OWM positions its API's, it feels that the _OneCall_ API offers the best 'value for money' and a lot of useful information. It needs just __1__ call (hence it being called _OneCall_ :) ) to retrieve all relevant information so the free limit of 1000 per day seems more than adequate (every 10 mins = 6 per hour X 24 hours = 144 per day!). 

The _Professional_ subscription on the other hand, gives much more flexibility and some more options like the _Geo_ API, WeatherMaps, WeatherTriggers but also Widgets that could be leveraged in the UI, etc. It would make it possible to create a much more advanced Weather module in Domoticz.

Or maybe we need to split things into 2 independent OWM modules. 1 'simple' one using _OneCall_ API (and its subscription) and an 'advanced' module leveraging the _Professional_ subscription (and levels).

Let's continue this discussion on the Domoticz Slack channel(s) and/or forum.

**Comment by gsagsa99** [comment]
Thanks

What is next ?.  As new customer (and I guess I will not be alone), I am
blocked with this issue.

regards


[image: width=]
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

Le mer. 14 sept. 2022 à 21:26, KidDigital ***@***.***> a
écrit :

**Comment by kiddigital** [comment]
> Thanks
> 
> What is next ?.  As new customer (and I guess I will not be alone), I am
> blocked with this issue.
> 
> regards
> 
> 

Yes, unfortunately you are. 

But I think you agree changing the existing module now, will cause that existing users functionality becomes broken. That is not an option.

We need to decide a way forward. And find people/time to build it.

> >
> > Let's continue this discussion on the Domoticz Slack channel(s) and/or
> > forum.
> >

**Comment by gsagsa99** [comment]
Fine with me

As I am really new in this, what do you mean 
"
Let's continue this discussion on the Domoticz Slack channel(s) and/or forum "

Which channel .  Shall I open an item in the forum ?  Which one ?

Sorry but this is the first time for me .
regards

**Comment by gsagsa99** [comment]
Hi @kiddigital 

I don't know if it is feasible and I don't know the amount of work if would be to a get a  'dedicate' version with version 3.0 I could use ?

Which code need to be recompiled ?

Regards

**Comment by gsagsa99** [comment]
Hi

a better solution would be to have the version as a parameter , default being 2.5.  Feasible ?

regards

**Comment by kiddigital** [comment]
First have to test if OneCall 2.5 (actually OWM calls it 1.0) has the same result structure as 3.0.

If those are the same, than creating a 3.0 'version' is relatively easy.

If people have time and ability to test, that would be great. I will do the same when time allows.

@gsagsa99 , could you attach a full output of a call so maybe I can do the same call against 2.5 and run a diff?

**Comment by gsagsa99** [comment]
Hello @kiddigital 

Please find attached, the query, the response as well as the formatted json.

I you need to change the query (additional params) let me know, 

When a "modified version" is available, I am ready to test it even if I am not expert :-)

[Full.zip](https://github.com/domoticz/domoticz/files/9595027/Full.zip)

**Comment by gsagsa99** [comment]
hello @kiddigital 

I don't know what is going on at OWM site but the good new is that the  oncall API 2.5 is WORKING since 23 sep 2022.

As you can see in the following log , during the same day  
1. Morning the call was failing (around 8:22)
2. In the afternoon the same call was OK (arounf 17:51)

2022-09-23 08:22:53.794  Debug: Meteo_Local: Get data from https://api.openweathermap.org/data/2.5/onecall?lat=48.8565&lon=2.3524&exclude=minutely&appid=f4bc1306c8c0bda6307d0dc8941437a4&units=metric&lang=fr
2022-09-23 08:22:53.905  Debug: HTTP 401: Unauthorized
2022-09-23 08:22:53.905  Error: Meteo_Local: Error getting http data!


2022-09-23 17:51:59.306  Debug: Meteo_Local: Get data from https://api.openweathermap.org/data/2.5/onecall?lat=48.8565&lon=2.3524&exclude=minutely&appid=f4bc1306c8c0bda6307d0dc8941437a4&units=metric&lang=fr
2022-09-23 17:51:59.580  Meteo_Local: General/Text (Weather Description)
2022-09-23 17:51:59.587  Meteo_Local: Temp + Humidity + Baro (TempHumBaro)
2022-09-23 17:51:59.592  Meteo_Local: Temp (Feel Temperature)
2022-09-23 17:51:59.597  Meteo_Local: Wind (Wind)
2022-09-23 17:51:59.601  Meteo_Local: General/Visibility (Visibility)
2022-09-23 17:51:59.601  Meteo_Local: General/Percentage (Clouds %)
2022-09-23 17:51:59.601  Meteo_Local: Rain (Precipitation)
2022-09-23 17:51:59.601  Meteo_Local: Lighting 2 (Is it raining/snowing)



of course the setup did not change in Domoticz.  I did not check at the OWM yet to see if something happened:  Rollback to previous API ?

The API key I am using in my setup is 'Base Plan'

If I use the API key free (no plan) , it is still failed.

In summary

1. New account with Base plan is OK, 
https://api.openweathermap.org/data/2.5/onecall?lat=50.7001&lon=4.988987&appid=f4bc1306c8c0bda6307d0dc8941437a4&units=metric&lang=fr ---> OK

2. New account no Base plan is FAILING 
https://api.openweathermap.org/data/2.5/onecall?lat=50.7001&lon=4.988987&appid=364d01e717bf215a28cc072aed7234cd&units=metric&lang=fr ---> FAILED

This is good new as I am blocked anymore

**Comment by gsagsa99** [comment]
Fine with me

As I am really new in this, what do you mean
"
Let's continue this discussion on the Domoticz Slack channel(s)
and/or forum "

Which chanel .  Shall I open an item in the forum ?  Which one ?

Sorry but this is the first time for me .
regards

**Comment by Zaphod-dev** [comment]
Hi, I can confirm that I was unable to display the OWM forecast using domoticz after creating a brand new "free" API key (with the same "Error getting http data" error). After subscribing to the "Base Plan" (and crossing fingers I never reach the 2k calls) I immediatly could display the forecast in domoticz (errors also stopped after a while). 
Indeed if the plugin could be made to cater for old & new OWM API keys (without the need for paid plans) it would be such a relief !
Kind regards

**Comment by gizmocuz** [comment]
@gsagsa99 , maybe create a channel OpenWeatherMap ?

**Comment by gizmocuz** [comment]
As you can see in the code, the API is called every 300 seconds = every 5 minutes
that is 288 calls a day, is 9000 calls a month
I think you should be safe with the free plan?

**Comment by pipiche38** [comment]
I have a similar issue, via Domoticz I'm unable to get it working and have : 

> Mar 23 09:07:44 pi3 domoticz[7582]: 2023-03-23 09:07:44.800  Status: OpenWeatherMap: Worker stopped...
> Mar 23 09:07:44 pi3 domoticz[7582]: 2023-03-23 09:07:44.802  Status: OpenWeatherMap: Using specified location (Lon lon=5.8333, Lat lat=45.0833)!
> Mar 23 09:07:44 pi3 domoticz[7582]: 2023-03-23 09:07:44.802  Status: OpenWeatherMap: Started
> Mar 23 09:07:48 pi3 domoticz[7582]: 2023-03-23 09:07:48.056  Error: OpenWeatherMap: Error getting http data!

calling the API with the URL + the right Key works
http://api.openweathermap.org/data/2.5/weather?q=London,uk&APPID=

**Comment by pipiche38** [comment]
Is there any chance to get this addressed ?

Same problem with Onecall v3., old keys still working.
https://www.domoticz.com/forum/viewtopic.php?p=301207&hilit=openweathermap#p301207

**Comment by kiddigital** [comment]
See the discussion at the beginnen of this issue.

We can not update the existing module as it will break functionality for users with an older API key.

So someone might have to develop a second OWM module that uses the new OneCall API that could be used by users that have the new API key and subscription.

Once I have to time, I might give it a go but happy if others beat me to it 😁

**Comment by gizmocuz** [comment]
Why not add a option on the hardware setup page to use V1 or V2, oe maybe it is possible to detect it?
This way we keep the code in the right source file

**Comment by patrickschaerer** [comment]
> Hi, I can confirm that I was unable to display the OWM forecast using domoticz after creating a brand new "free" API key (with the same "Error getting http data" error). After subscribing to the "Base Plan" (and crossing fingers I never reach the 2k calls) I immediatly could display the forecast in domoticz (errors also stopped after a while). Indeed if the plugin could be made to cater for old & new OWM API keys (without the need for paid plans) it would be such a relief ! Kind regards

this worked for me.
Thanks

**Comment by pipiche38** [comment]
> Why not add a option on the hardware setup page to use V1 or V2, oe maybe it is possible to detect it? This way we keep the code in the right source file

Alternative could be:

e5e3ee5894b15cee562c06cc9735686.  -> interpreted as a 2.5 API
V2.5:e5e3ee5894b15cee562c06cc9735686 -> interpreted as 2.5 API
V3.0:e5e3ee5894b15cee562c06cc9735686 -> interpreted as 3.0 API

**Comment by pipiche38** [comment]
I'm suggesting a code like that

```
#define OWM_30_onecall_URL "https://api.openweathermap.org/data/3.0/onecall?"

    // Check API version
    if (m_APIKey.substr(0, 4) == "V3.0:")
    {
        sURL << OWM_30_onecall_URL;
        m_APIKey = m_APIKey.substr(4); // Remove the "V3.0:" prefix
    }
    // Check if m_APIKey starts with "V2.5:"
    else if (m_APIKey.substr(0, 4) == "V2.5:")
    {
        sURL << OWM_25_onecall_URL;
        m_APIKey = m_APIKey.substr(4); // Remove the "V2.5:" prefix
    }
    else
    {
        sURL << OWM_onecall_URL;
    }

    sURL << "lat=" << m_Lat << "&lon=" << m_Lon;
    sURL << "&exclude=minutely";
    sURL << "&appid=" << m_APIKey;
    sURL << "&units=metric" << "&lang=" << m_Language;

```

**Comment by pipiche38** [comment]
@gizmocuz see #6053 I did my best to fix this. Hope that is convenient for everyone

**Comment by gizmocuz** [comment]
@pipiche38 , thanks for this!!! Aldo this works, it might be better to make a dropdown to select the API and put it in one of the free 'mode' fields (0=2.5=default, 1=3.0)
Is it possible to change it like this?

**Comment by pipiche38** [comment]
then better to find someone who can handle the web part. 
I did my best here, sorry

**Comment by gizmocuz** [comment]
OpenWeatherMap API 3.0 is now supported in beta 15932.
You can specify the API version to use in the web interface.
For existing setup this is set to 2.5, but for new setup this is set to 3.0 by default
